contextual_lenses/encoders.py

```diff
@@ -0,0 +1,157 @@
+"""Encoder functions
+
+Fixed and learnable transformations for embedding sequences."""
+
+
+import flax
+from flax import nn
+
+import jax
+from jax import lax
+import jax.nn
+import jax.numpy as jnp
+
+import numpy as np
+
+from operator import itemgetter
+
+
+def one_hot_encoder(batch_inds, num_categories):
+  """Applies one-hot encoding from jax.nn."""
+
+  one_hots = jax.nn.one_hot(batch_inds, num_classes=num_categories)
+  
+  return one_hots
+
+
+class CNN(nn.Module):
+  """A simple 1D CNN model."""
+
+  def apply(self, x, n_layers, n_features, n_kernel_sizes):
+    
+    x = jnp.expand_dims(x, axis=2)
+
+    for layer in range(n_layers):
+      features = n_features[layer]
+      kernel_size = (n_kernel_sizes[layer], 1)
+      x = nn.Conv(x, features=features, kernel_size=kernel_size)
+      x = nn.relu(x)
+    
+    x = jnp.squeeze(x, axis=2)
+
+    return x
+
+
+def cnn_one_hot_encoder(batch_inds, num_categories, n_layers, n_features, n_kernel_sizes):
+  """Applies one-hot encoding followed by 1D CNN."""
+
+  one_hots = one_hot_encoder(batch_inds, num_categories)
+  cnn_one_hots = CNN(one_hots, n_layers, n_features, n_kernel_sizes)
+  
+  return cnn_one_hots
+
+
+"""Positional embeddings
+Code source: https://github.com/google/flax/blob/aff10f032e892e28a1acf4dd4ee9dcc6cd39a606/examples/wmt/models.py."""
```

Comment:
style: use # to add comments in middle of file. Generally only use """ for docstrings and the one at top of files.